vm/src/protocol/number.rs

```diff
@@ -1,189 +1,155 @@
+        true_divide: AtomicCell::new(None),
+        inplace_floor_divide: AtomicCell::new(None),
+        inplace_true_divide: AtomicCell::new(None),
+        index: AtomicCell::new(None),
+        matrix_multiply: AtomicCell::new(None),
+        inplace_matrix_multiply: AtomicCell::new(None),
     };
-
-    pub(crate) fn generic(
-        has_int: bool,
-        has_float: bool,
-        has_index: bool,
-    ) -> &'static PyNumberMethods {
-        static METHODS: &[PyNumberMethods] = &[
-            new_generic(false, false, false),
-            new_generic(true, false, false),
-            new_generic(false, true, false),
-            new_generic(true, true, false),
-            new_generic(false, false, true),
-            new_generic(true, false, true),
-            new_generic(false, true, true),
-            new_generic(true, true, true),
-        ];
-
-        fn int(num: &PyNumber, vm: &VirtualMachine) -> PyResult<PyRef<PyInt>> {
-            let ret = vm.call_special_method(num.obj.to_owned(), identifier!(vm, __int__), ())?;
-            ret.downcast::<PyInt>().map_err(|obj| {
-                vm.new_type_error(format!("__int__ returned non-int (type {})", obj.class()))
-            })
-        }
-        fn float(num: &PyNumber, vm: &VirtualMachine) -> PyResult<PyRef<PyFloat>> {
-            let ret = vm.call_special_method(num.obj.to_owned(), identifier!(vm, __float__), ())?;
-            ret.downcast::<PyFloat>().map_err(|obj| {
-                vm.new_type_error(format!(
-                    "__float__ returned non-float (type {})",
-                    obj.class()
-                ))
-            })
-        }
-        fn index(num: &PyNumber, vm: &VirtualMachine) -> PyResult<PyRef<PyInt>> {
-            let ret = vm.call_special_method(num.obj.to_owned(), identifier!(vm, __index__), ())?;
-            ret.downcast::<PyInt>().map_err(|obj| {
-                vm.new_type_error(format!("__index__ returned non-int (type {})", obj.class()))
-            })
-        }
-
-        const fn new_generic(has_int: bool, has_float: bool, has_index: bool) -> PyNumberMethods {
-            PyNumberMethods {
-                int: if has_int { Some(int) } else { None },
-                float: if has_float { Some(float) } else { None },
-                index: if has_index { Some(index) } else { None },
-                ..PyNumberMethods::NOT_IMPLEMENTED
-            }
-        }
-
-        let key = (has_int as usize) | ((has_float as usize) << 1) | ((has_index as usize) << 2);
-
-        &METHODS[key]
-    }
 }
 
 pub struct PyNumber<'a> {
     pub obj: &'a PyObject,
     // some fast path do not need methods, so we do lazy initialize
-    pub methods: Option<&'static PyNumberMethods>,
+    methods: OnceCell<PointerSlot<PyNumberMethods>>,
 }
 
-impl<'a> PyNumber<'a> {
-    pub fn new(obj: &'a PyObject, vm: &VirtualMachine) -> Self {
+impl<'a> From<&'a PyObject> for PyNumber<'a> {
+    fn from(obj: &'a PyObject) -> Self {
```

Comment:
This change means having a `PyNumber` object doesn't guarantee it is an actual number object.
Do you intend to change it?
In that case, we'd better to carefully comment about it because it is counter-intutitive to rust sense.

I still think `OnceCell` for `methods` is an overkill because it can be safely managed only with properly designed types.

Reply:
I think PyNumber shouldn't garrentee anything as it could be int, float, subclass, complex or class with __int__, __float__ etc, so what the caller expect is depend on what method been called.

The reason I use OnceCell is because we always use the type id first to avoid method call and I think that is more common case like get int from PyInt

Reply:
We will obtain an `PyNumber` object only when we need `PyNumber` operations, isn't it? The method can be obtained when we create `PyNumber`, not when retrieving methods from it. `PyNumber` itself is a `Once`. `OnceCell` in `PyNumber` is duplicated encoding of data.

Reply:
so maybe a function that will create PyNumber only if it is not exact type could be better?

Reply:
I am sorry. I didn't understand well what you meant by "create PyNumber only if it is not exact type". could you help me to understand it?

I think we need to create a `PyNumber` type object only when we need to call `methods()` of it. In that case, we don't need `OnceCell` for PyNumber - because we will create a `PyNumber` object always with methods. If we don't need method, we can handle `PyObjectRef` instead of `PyNumber`.

Reply:
because recent patches are adding number protocol implementations, quickly merging this one will be less painful in future.
I want to know how do you think about this OnceCell problem for now. You know about it the best.

Reply:
@qingshi163 do you mind if I try to edit this PR to work this way? I want to listen your decision because you are the best professional for PyNumber.

Reply:
I am thinking we should remove OnceCell from PyNumber but keep From<&PyObject> trait.
Than we modify or create function under VirtualMachine(to_int, to_index, to_float, etc) for all the function that could short cut with type id and do the check there. so we will always call number methods if we created PyNumber.

We should always use functions under VirtualMachine instead PyNumber if we do not sure the type.

I'd like to keep From<&PyObject> is because PyNumber is not garuentee anything about the type. we should not check the type when creating it.

Reply:
Sure you can modify the pr, I am quite busy and I want to fix the sre-engine first.